### ISSUE TITLE
How could leela zero use TPU for training?

### ISSUE THREAD

**Issue by ionicliquids11** [tensorflow/minigo#541]
Dear Sir/Madam

Is there any method for Leela Zero(https://github.com/gcp/leela-zero) training to use colab's free TPU? 

Maybe I shouldn't ask this question here, but I know it is the only place that I can get the answer since only here have the experts who know TPU  and also love go AI. 

I am looking forwards to your any kind reply~

Thanks a lot!

Ionic liquids

**Comment by sethtroisi** [comment]
TL;DR It would take a lot of code for moderate gains because of LZ's project setup

I would guess the easiest way to support TPUs would be to extend the [IPC (InterProcess Communication) that NTT123 (and later I) wrote](https://github.com/sethtroisi/leela-zero/tree/y_udp_nn/ipc). We basically added some code to [Network.cpp](https://github.com/sethtroisi/leela-zero/blob/y_udp_nn/src/Network.cpp#L608) that pushes the NN inputs into shared memory. We then read them from a python process and evaluated them with theano. At that point you'd read the values in python (similar to this projects old [inference_worker.py](https://github.com/tensorflow/minigo/blob/a6fb1a2f74f6ec835366649f081963745f8f9160/inference_worker.py)) and call tensorflow run. This is glossing over a lot of details.

1. Modify Network to pass weights via grpc or shared memory to python [Easy]
1. Convert [tfprocess.py](https://github.com/gcp/leela-zero/blob/master/training/tf/tfprocess.py) to tf.estimator [Medium]
1. Ask GCP to distribute tensorflow weights for write ascii to tf weights converter. [Hard]
1. Convert inference_worker to use tfprocess.py model and read from shared memory [Easy]
1. Rewrite some parts of gtp so that all games use the same model.* [Easy]

This would get the code running on TPUs but to see the real speed of TPUs you'd need much larger batches of positions. This means doing virtual loss (This was partially supported in the past I believe)

1. Figure out a good virtual loss setup [Hard]



\* Figure out what to do with 10-20 partial games if the model changes

**Comment by amj** [comment]
@ionicliquids11 if you need more info, feel free to re-open!

**Comment by alreadydone** [comment]
From https://arxiv.org/abs/1902.04522 Appendix A, ELF OpenGo doesn't worry about "partial games if the model changes" at all:
> Asynchronized (AlphaZero) mode 
Note that AlphaGo Zero mode involves a lot of synchronization and is not efficient in terms of boosting the strength of the trained model. In AlphaZero mode, we release all the synchronization locks and remove Eval clients. Next moves are always generated using the latest models and Selfplay clients does not terminate their current games upon new models. It is possible that for a given selfplay game record, the first part of the game is played by model A while the second part of the game is played by model B.
We initially started with the synchronous approach before switching to the asynchronous approach. Switching offered two benefits: (1) Both selfplay generation and training realized a drastic speedup. The asynchronous approach achieves over 5x the selfplay throughput of the synchronous approach on our hardware setup. (2) The ratio of selfplay games to training minibatches increased by , thus helping to prevent overfitting. 
The cost of the asynchronous approach is homogeneity of selfplays – each selfplay is now the product of many consecutive models, reducing the internal coherency of the moves. However, we note that the replay buffer that provides training data is already extremely heterogenous, typically containing games from over 25 different models. Consequently, we suspect and have empirically verified that the effect of within-selfplay heterogeneity is mild.
